From: Shaohua Li <shaohua.li@intel.com>
Date: Fri, 4 Dec 2009 13:12:06 +0100
Subject: [PATCH] cfq-iosched: make nonrot check logic consistent
Git-commit: 3c764b7a654668dd04905841d6024f7b6aa843a5
References: FATE#311054
Patch-Mainline: 2.6.33

cfq_arm_slice_timer() has logic to disable idle window for SSD device. The same
thing should be done at cfq_select_queue() too, otherwise we will still see
idle window. This makes the nonrot check logic consistent in cfq.
Tests in a intel SSD with low_latency knob close, below patch can triple disk
thoughput for muti-thread sequential read.

Signed-off-by: Shaohua Li <shaohua.li@intel.com>
Signed-off-by: Jens Axboe <jens.axboe@oracle.com>
Signed-off-by: Hannes Reinecke <hare@suse.de>
---
 block/cfq-iosched.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/block/cfq-iosched.c b/block/cfq-iosched.c
index d338016..63fd146 100644
--- a/block/cfq-iosched.c
+++ b/block/cfq-iosched.c
@@ -1798,7 +1798,8 @@ static bool cfq_should_idle(struct cfq_data *cfqd, struct cfq_queue *cfqq)
 		return false;
 
 	/* We do for queues that were marked with idle window flag. */
-	if (cfq_cfqq_idle_window(cfqq))
+	if (cfq_cfqq_idle_window(cfqq) &&
+	   !(blk_queue_nonrot(cfqd->queue) && cfqd->hw_tag))
 		return true;
 
 	/*
-- 
1.6.0.2

